src/helpers/query-daily-data.tsx

```diff
@@ -0,0 +1,94 @@
+﻿import { add } from "date-fns";
+import getDayKey from "./get-day-key";
+import { DailyDataType, DailyDataTypeDefinition } from "./daily-data-types";
+import allTypeDefinitions from "./daily-data-types/all";
+import { FontAwesomeSvgIcon } from "react-fontawesome-svg-icon";
+import { faPersonRunning } from "@fortawesome/free-solid-svg-icons";
+import React from "react";
+import { defaultFormatter } from "./daily-data-types/formatters";
+
+export type DailyDataQueryResult = { [key: string]: number };
+export type DailyDataProvider = (startDate: Date, endDate: Date) => Promise<DailyDataQueryResult>;
+export type DailyDataAvailabilityCheck = (modifiedAfter?: Date) => Promise<boolean>;
+
+let dailyDataTypes = new Map(
+	allTypeDefinitions.map((typeDefinition) => [typeDefinition.type, typeDefinition])
+);;
+
+export function registerDailyDataProvider(type: string, provider: DailyDataProvider, availabilityCheck: DailyDataAvailabilityCheck) {
+	dailyDataTypes.set(type, {
+		type: type,
+		getLabel: () => type,
+		dataProvider: provider,
+		availabilityCheck: availabilityCheck,
+		icon: <FontAwesomeSvgIcon icon={faPersonRunning} />,
+		formatter: defaultFormatter,
+		previewDataRange: [0, 10000]
+	});
+}
+
+export function registerDailyDataTypeDefinition(typeDefinition: DailyDataTypeDefinition) {
+	dailyDataTypes.set(typeDefinition.type, typeDefinition);
+}
+
+export function checkDailyDataAvailability(type: string, modifiedAfter?: Date) {
+	var dailyDataType = dailyDataTypes.get(type);
+	if (!dailyDataType) { throw "Unknown data type:" + type; }
+	return dailyDataType.availabilityCheck(modifiedAfter);
+}
+
+export function queryDailyData(type: string, startDate: Date, endDate: Date, preview?: boolean) {
+	if (preview) {
+		return Promise.resolve(queryPreviewDailyData(type, startDate, endDate));
+	}
+	var dailyDataType = dailyDataTypes.get(type);
+	if (!dailyDataType) { throw "Unknown data type:" + type; }
+	return dailyDataType.dataProvider(startDate, endDate).then(function (data) {
+		var result: DailyDataQueryResult = {};
+		while (startDate < endDate) {
+			var dayKey = getDayKey(startDate);
+			if (data[dayKey]) {
+				result[dayKey] = data[dayKey];
+			}
+			startDate = add(startDate, { days: 1 });
+		}
+		return result;
+	});
+}
+
+export async function queryPreviewDailyData(type: string, startDate: Date, endDate: Date) {
+	var result: DailyDataQueryResult = {};
+	let range = getDailyDataTypeDefinition(type as DailyDataType).previewDataRange;
+
+	//poor man's seeded javascript rng
+	//just for generating a relatively large random integer from a string seed,
+	//which then uses modulo to get a number in a range
+	async function randomNumber(message: string) {
```

Comment:
Why go through all this trouble for a preview state random number?  Why not just use `Math.random()`?

Reply:
Math.random() can't be seeded so can't be made deterministic.   See PR notes for why I wanted a deterministic solution.

Couple reasons:
- Randomly generating different data on each render means the graphs get very "jumpy" under certain circumstances, like in view builder in particular
- It can make it hard to tell a consistent story if you're building a flowboard or something; like lets say we have a graph of steps that shows 5,000 for 5/1, then below that a component that shows 3,000 for 5/1 because they were both generating their own random data
  - Originally I wasn't planning on passing the data from RelativeActivityCoordinator to the RelativeActivity component, so this was very noticeable there since the bars at the top for the selected day wouldn't match the bars for the day...

But yeah, this solution basically means "Steps" for 5/1 will always be the same number regardless of where it's showing

Reply:
Makes sense.  Thanks for the explanation.